File: ios/Classes/controller/overlay/handler/LabelControllerHandler.swift
import KakaoMapsSDK
import Flutter

internal protocol LabelControllerHandler {
    var labelManager: LabelManager { get };

    func createLabelLayer(option: LabelLayerOptions, onSuccess: (Any?) -> Void)

    func removeLabelLayer(layerId: String, onSuccess: (Any?) -> Void)

    func addPoiStyle(style: PoiStyle, onSuccess: (String) -> Void)

    func addPoi(layer: LabelLayer, poi: PoiOptions, position: MapPoint, visible: Bool, onSuccess: @escaping (String) -> Void)

    func removePoi(layer: LabelLayer, poiId: String, onSuccess: (Any?) -> Void)

    func addPolylineText(layer: LabelLayer, label: WaveTextOptions, visible: Bool, onSuccess: (String) -> Void)

    func removePolylineText(layer: LabelLayer, labelId: String, onSuccess: (Any?) -> Void)

    func changePoiPixelOffset(poi: Poi, offset: CGPoint, onSuccess: (Any?) -> Void)

    func changePoiVisible(poi: Poi, visible: Bool, onSuccess: (Any?) -> Void)

    func changePoiStyle(poi: Poi, styleId: String, transition: Bool, onSuccess: (Any?) -> Void)

    func changePoiText(poi: Poi, styleId: String, text: String, transition: Bool, onSuccess: (Any?) -> Void)

    func invalidatePoi(
        poi: Poi,
        styleId: String,
        text: String,
        transition: Bool,
        onSuccess: (Any?) -> Void
    )
    
    func movePoi(poi: Poi, position: MapPoint, duration: UInt?, onSuccess: (Any?) -> Void)

    func rotatePoi(poi: Poi, angle: Double, duration: UInt?, onSuccess: (Any?) -> Void)

    func rankPoi(poi: Poi, rank: Int, onSuccess: (Any?) -> Void)
}

internal extension LabelControllerHandler {
    func labelHandle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = castSafty(call.arguments, caster: asDict)
        let layerId: String? = castSafty(arguments?["layerId"], caster: asString)
        let layer: LabelLayer? = layerId.flatMap { key in
            labelManager.getLabelLayer(layerID: key)
        }

        let poiId = castSafty(arguments?["poiId"], caster: asString)
        let poi: Poi? = poiId.flatMap { key in
            layer!.getPoi(poiID: key)
        }

        let polylineTextId = castSafty(arguments?["labelId"], caster: asString)
        let polylineText: WaveText? = polylineTextId.flatMap { key in
            layer!.getWaveText(waveTextID: key)
        }

        switch call.method {
        case "createLabelLayer": createLabelLayer(option: LabelLayerOptions(payload: arguments!), onSuccess: result)
        case "removeLabelLayer": removeLabelLayer(layerId: layerId!, onSuccess: result)
        case "addPoiStyle": addPoiStyle(style: PoiStyle(payload: arguments!), onSuccess: result)
        case "addPoi":
            let poiArgument = asDict(arguments!["poi"]!)
            let poiOption = PoiOptions(payload: poiArgument)
            let position = MapPoint(payload: poiArgument)
            let visible = asBool(arguments!["visible"] ?? true)
            addPoi(layer: layer!, poi: poiOption, position: position, visible: visible, onSuccess: result)
        case "removePoi": removePoi(layer: layer!, poiId: poiId!, onSuccess: result)
        case "addPolylineText":
            let waveTextArgument = asDict(arguments!["label"]!)
            let waveTextStyle = WaveTextStyle(payload: asDict(waveTextArgument["style"]!))
            labelManager.addWaveTextStyle(waveTextStyle)
            let waveTextOption = WaveTextOptions(payload: waveTextArgument, styleId: waveTextStyle.styleID)
            let visible = asBool(arguments!["visible"] ?? true)
            addPolylineText(layer: layer!, label: waveTextOption, visible: visible, onSuccess: result)
        case "removePolylineText": removePolylineText(layer: layer!, labelId: poiId!, onSuccess: result)
        // poi Handler
        case "changePoiPixelOffset":
            let rawPayload: Dictionary<String, Double> = ["x": asDouble(arguments!["x"]!), "y": asDouble(arguments!["y"]!)]
            let offset = CGPoint(payload: rawPayload)
            changePoiPixelOffset(poi: poi!, offset: offset, onSuccess: result)
        case "changePoiVisible":
            let visible = asBool(arguments!["visible"]!)
            changePoiVisible(poi: poi!, visible: visible, onSuccess: result)
        case "changePoiStyle":
            let styleId = asString(arguments!["styleId"]!)
            let transition = asBool(arguments!["transition"] ?? false)
            changePoiStyle(poi: poi!, styleId: styleId, transition: transition, onSuccess: result)
        case "changePoiText":
            let text = asString(arguments!["text"]!)
            let transition = asBool(arguments!["transition"] ?? false)
            let styleId = asString(arguments!["styleId"]!)
            changePoiText(poi: poi!, styleId: styleId, text: text, transition: transition, onSuccess: result)
        case "invalidatePoi":
            let styleId = asString(arguments!["styleId"]!)
            let text = asString(arguments!["text"]!)
            let transition = asBool(arguments!["transition"] ?? false)
            invalidatePoi(poi: poi!, styleId: styleId, text: text, transition: transition, onSuccess: result)
        case "movePoi":
            let position = MapPoint(payload: arguments!)
            let duration = castSafty(arguments!["millis"], caster: asUInt)
            movePoi(poi: poi!, position: position, duration: duration, onSuccess: result)
        case "rotatePoi":
            let angle = asDouble(arguments!["angle"]!)
            let duration = castSafty(arguments!["millis"], caster: asUInt)
            rotatePoi(poi: poi!, angle: angle, duration: duration, onSuccess: result)
        case "rankPoi":
            let rank = asInt(arguments!["rank"]!)
            rankPoi(poi: poi!, rank: rank, onSuccess: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/kr/yhs/flutter_kakao_maps/controller/overlay/handler/LabelControllerHandler.kt
package kr.yhs.flutter_kakao_maps.controller.overlay.handler

import com.kakao.vectormap.LatLng
import com.kakao.vectormap.label.Label
import com.kakao.vectormap.label.LabelLayer
import com.kakao.vectormap.label.LabelLayerOptions
import com.kakao.vectormap.label.LabelManager
import com.kakao.vectormap.label.LabelOptions
import com.kakao.vectormap.label.LabelStyles
import com.kakao.vectormap.label.PolylineLabel
import com.kakao.vectormap.label.PolylineLabelOptions
import com.kakao.vectormap.label.PolylineLabelStyles
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kr.yhs.flutter_kakao_maps.converter.CameraTypeConverter.asLatLng
import kr.yhs.flutter_kakao_maps.converter.LabelTypeConverter.asLabelLayerOptions
import kr.yhs.flutter_kakao_maps.converter.LabelTypeConverter.asLabelOptions
import kr.yhs.flutter_kakao_maps.converter.LabelTypeConverter.asLabelStyles
import kr.yhs.flutter_kakao_maps.converter.LabelTypeConverter.asPolylineTextOption
import kr.yhs.flutter_kakao_maps.converter.LabelTypeConverter.asPolylineTextStyles
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asBoolean
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asFloat
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asInt
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asLong
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asMap
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asString

interface LabelControllerHandler {
    val labelManager: LabelManager?

    fun labelHandle(call: MethodCall, result: MethodChannel.Result) {
        val arguments = call.arguments!!.asMap<Any?>()
        if (labelManager == null) {
            throw NullPointerException("LabelManager is null.")
        }

        val layer =
                arguments["layerId"]?.asString()?.let<String, LabelLayer> {
                    labelManager!!.getLayer(it)
                }
        val poi = layer?.run { arguments["poiId"]?.asString()?.let(layer::getLabel) }
        val polylineText =
                layer?.run { arguments["labelId"]?.asString()?.let(layer::getPolylineLabel) }

        when (call.method) {
            "createLabelLayer" -> {
                createLabelLayer(arguments.asLabelLayerOptions(), result::success)
            }
            "removeLabelLayer" -> {
                removeLabelLayer(layer!!, result::success)
            }
            "addPoiStyle" -> {
                addPoiStyle(arguments.asLabelStyles()!!, result::success)
            }
            "addPoi" -> {
                val poiOption = arguments["poi"]!!.asLabelOptions(labelManager!!)
                addPoi(layer!!, poiOption, result::success)
            }
            "addPolylineText" -> {
                val labelOption = arguments["label"]!!.asPolylineTextOption()
                addPolylineText(layer!!, labelOption, result::success)
            }
            "removePoi" -> removePoi(layer!!, poi!!, result::success)
            "removePolylineText" -> removePolylineText(layer!!, polylineText!!, result::success)
            // poi Handler
            "changePoiOffsetPosition" -> {
                val x = arguments["x"]?.asFloat()!!
                val y = arguments["y"]?.asFloat()!!
                val forceDpScale = arguments["forceDpScale"]?.asBoolean()
                changePoiOffsetPosition(poi!!, x, y, forceDpScale, result::success)
            }
            "changePoiVisible" -> {
                val visible = arguments["visible"]?.asBoolean()!!
                changePoiVisible(poi!!, visible, result::success)
            }
            "changePoiStyle" -> {
                val styleId = arguments["styleId"]?.asString()!!
                val transition = arguments["transition"]?.asBoolean() ?: false
                changePoiStyle(poi!!, styleId, transition, result::success)
            }
            "changePoiText" -> {
                val text = arguments["text"]?.asString()!!
                val transition = arguments["transition"]?.asBoolean() ?: false
                changePoiText(poi!!, text, transition, result::success)
            }
            "invalidatePoi" -> {
                val styleId = arguments["styleId"]?.asString()!!
                val text = arguments["text"]?.asString()!!
                val transition = arguments["transition"]?.asBoolean() ?: false
                invalidatePoi(poi!!, styleId, text, transition, result::success)
            }
            "movePoi" -> {
                val position = arguments.asLatLng()
                val millis = arguments["millis"]?.asInt()
                movePoi(poi!!, position, millis, result::success)
            }
            "rotatePoi" -> {
                val angle = arguments["angle"]?.asFloat()!!
                val millis = arguments["millis"]?.asInt()
                rotatePoi(poi!!, angle, millis, result::success)
            }
            "scalePoi" -> {
                val x = arguments["x"]?.asFloat()!!
                val y = arguments["y"]?.asFloat()!!
                val millis = arguments["millis"]?.asInt()
                scalePoi(poi!!, x, y, millis, result::success)
            }
            "rankPoi" -> {
                val rank = arguments["rank"]?.asLong()!!
                rankPoi(poi!!, rank, result::success)
            }
            // polyline text handler
            "changePolylineTextStyle" -> {
                val style = arguments["styles"]?.asPolylineTextStyles()
                val text = arguments["text"]?.asString()
                changePolylineTextAndStyle(polylineText!!, style!!, text, result::success)
            }
            "changePolylineTextVisible" -> {
                val visible = arguments["visible"]?.asBoolean()!!
                changePolylineTextVisible(polylineText!!, visible, result::success)
            }
            else -> result.notImplemented()
        }
    }

    fun createLabelLayer(options: LabelLayerOptions, onSuccess: (Any?) -> Unit)

    fun removeLabelLayer(layer: LabelLayer, onSuccess: (Any?) -> Unit)

    fun addPoiStyle(style: LabelStyles, onSuccess: (Any?) -> Unit)

    fun addPoi(layer: LabelLayer, poi: LabelOptions, onSuccess: (String) -> Unit)

    fun removePoi(layer: LabelLayer, poi: Label, onSuccess: (Any?) -> Unit)

    fun addPolylineText(layer: LabelLayer, label: PolylineLabelOptions, onSuccess: (String) -> Unit)

    fun removePolylineText(layer: LabelLayer, label: PolylineLabel, onSuccess: (Any?) -> Unit)

    // Poi Controller
    fun changePoiOffsetPosition(
            poi: Label,
            x: Float,
            y: Float,
            forceDpScale: Boolean?,
            onSuccess: (Any?) -> Unit
    )

    fun changePoiVisible(poi: Label, visible: Boolean, autoMove: Boolean?, duration: Int?, onSuccess: (Any?) -> Unit)

    fun changePoiStyle(poi: Label, styleId: String, transition: Boolean, onSuccess: (Any?) -> Unit)

    fun changePoiText(poi: Label, text: String, transition: Boolean, onSuccess: (Any?) -> Unit)

    fun invalidatePoi(
            poi: Label,
            styleId: String,
            text: String,
            transition: Boolean,
            onSuccess: (Any?) -> Unit
    )

    fun movePoi(poi: Label, position: LatLng, millis: Int?, onSuccess: (Any?) -> Unit)

    fun rotatePoi(poi: Label, angle: Float, millis: Int?, onSuccess: (Any?) -> Unit)

    fun scalePoi(poi: Label, x: Float, y: Float, millis: Int?, onSuccess: (Any?) -> Unit)

    fun rankPoi(poi: Label, rank: Long, onSuccess: (Any?) -> Unit)

    // Polyline Text Controller
    fun changePolylineTextAndStyle(
            label: PolylineLabel,
            style: PolylineLabelStyles,
            text: String?,
            onSuccess: (Any?) -> Unit
    )

    fun changePolylineTextVisible(label: PolylineLabel, visible: Boolean, onSuccess: (Any?) -> Unit)
}
